private/pkg/oauth2/client_test.go

```diff
@@ -0,0 +1,310 @@
+			if pollingCount == 0 {
+				pollingCount++
+				return newJSONResponse(t, http.StatusBadRequest, `{"error":"authorization_pending","error_description":"authorization pending"}`), nil
+			}
+			return newJSONResponse(t, http.StatusOK, `{"access_token":"accessToken","token_type":"bearer","expires_in":100,"refresh_token":"refreshToken","scope":"scope"}`), nil
+		},
+		output: &DeviceAccessTokenResponse{
+			AccessToken:  "accessToken",
+			TokenType:    "bearer",
+			ExpiresIn:    100,
+			RefreshToken: "refreshToken",
+			Scope:        "scope",
+		},
+	}, {
+		name: "error",
+		input: &DeviceAccessTokenRequest{
+			ClientID:     "clientID",
+			ClientSecret: "clientSecret",
+			DeviceCode:   "deviceCode",
+			GrantType:    "urn:ietf:params:oauth:grant-type:device_code",
+		},
+		transport: func(t *testing.T, r *http.Request) (*http.Response, error) {
+			assertFormRequest(t, r, url.Values{"client_id": {"clientID"}, "client_secret": {"clientSecret"}, "device_code": {"deviceCode"}, "grant_type": {"urn:ietf:params:oauth:grant-type:device_code"}})
+			return newJSONResponse(t, http.StatusBadRequest, `{"error":"invalid_request","error_description":"invalid request"}`), nil
+		},
+		err: &Error{
+			ErrorCode:        ErrorCodeInvalidRequest,
+			ErrorDescription: "invalid request",
+		},
+	}, {
+		name: "expired token",
+		input: &DeviceAccessTokenRequest{
+			ClientID:     "clientID",
+			ClientSecret: "clientSecret",
+			DeviceCode:   "deviceCode",
+			GrantType:    "urn:ietf:params:oauth:grant-type:device_code",
+		},
+		transport: func(t *testing.T, r *http.Request) (*http.Response, error) {
+			return newJSONResponse(t, http.StatusBadRequest, `{"error":"expired_token","error_description":"token expired"}`), nil
+		},
+		err: &Error{
+			ErrorCode:        ErrorCodeExpiredToken,
+			ErrorDescription: "token expired",
+		},
+	}}
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+			c := NewClient("https://buf.build", &http.Client{
+				Transport: roundTripFunc(func(r *http.Request) (*http.Response, error) {
+					assert.Equal(t, r.Method, http.MethodPost)
+					assert.Equal(t, r.URL.Path, DeviceTokenPath)
+					assert.Equal(t, r.Header.Get("Content-Type"), "application/x-www-form-urlencoded")
+					assert.Equal(t, r.Header.Get("Accept"), "application/json")
+					return test.transport(t, r)
+				}),
+			})
+			output, err := c.AccessDeviceToken(ctx, 1, test.input)
+			assert.Equal(t, test.output, output)
+			assert.Equal(t, err, test.err)
+		})
+	}
+}
+
+type roundTripFunc func(r *http.Request) (*http.Response, error)
+
+func (s roundTripFunc) RoundTrip(r *http.Request) (*http.Response, error) {
+	return s(r)
+}
+
+func newJSONResponse(t *testing.T, statusCode int, body string) *http.Response {
```

Comment:
`testNewJSONResponse` (prefix all test functions with `test`, just for consistency with the rest of the codebase)